### ISSUE TITLE
Switch from cuda1 to cuda2 introduces regression in NCCL collectives

### ISSUE THREAD

**Issue by sogartar** [iree-org/iree#16535]
### What happened?

I noticed that the swap from cuda1 to cuda2 driver may have caused a regression in the collectives HAL implementation.
```
boian-nvidia-gpus:7003:7022 [0] group.cc:274 NCCL WARN Cuda failure 'operation not permitted'

boian-nvidia-gpus:7003:7022 [0] init.cc:1786 NCCL WARN Cuda failure 'operation not permitted'
[boian-nvidia-gpus:07003] *** Process received signal ***
[boian-nvidia-gpus:07003] Signal: Segmentation fault (11)
[boian-nvidia-gpus:07003] Signal code:  (128)
[boian-nvidia-gpus:07003] Failing at address: (nil)
```

On commit e31b5a2e0476f0b7d0cf1b32869509554044c2c7, just before switching to cuda2. I don't get an error.

On another system there is no error at all.

### Steps to reproduce your issue

Running the test
```
NCCL_DEBUG=WARN python tests/e2e/collectives/collectives_test.py --target_backend=cuda --driver=cuda --verbose -k SingleRank.test_stablehlo_all_reduce
```

### What component(s) does this issue relate to?

Runtime

### Version information

4a80ee36d52b35be25bf30ea6a0f9b423477f6e9

### Additional context

Configure command:
```bash
export BUILD_TYPE=Debug

cmake \
  -GNinja \
  -B "$SCRIPT_DIR/build/ninja/$BUILD_TYPE" \
  -S "$SCRIPT_DIR/repo" \
  -DCMAKE_BUILD_TYPE=$BUILD_TYPE \
  -DCMAKE_EXPORT_COMPILE_COMMANDS=ON \
  -DIREE_ENABLE_ASSERTIONS=ON \
  -DLLVM_INCLUDE_EXAMPLES=OFF \
  -DLLVM_ENABLE_ASSERTIONS=ON \
  -DMLIR_INCLUDE_TESTS=ON \
  -DMLIR_INCLUDE_INTEGRATION_TESTS=OFF \
  -DIREE_BUILD_PYTHON_BINDINGS=ON \
  -DCMAKE_C_COMPILER=clang-15 \
  -DCMAKE_CXX_COMPILER=clang++-15 \
  -DPython3_EXECUTABLE="$(which python)" \
  -DIREE_TARGET_BACKEND_VULKAN_SPIRV=ON \
  -DIREE_TARGET_BACKEND_LLVM_CPU=ON \
  -DIREE_TARGET_BACKEND_CUDA=ON \
  -DIREE_HAL_DRIVER_CUDA=ON \
  -DIREE_ENABLE_LLD=ON \
  -DIREE_ENABLE_CCACHE=OFF \
  -DCMAKE_CXX_FLAGS=-gdwarf-4 \
  -DCMAKE_C_FLAGS=-gdwarf-4 \
  -DIREE_BUILD_EXPERIMENTAL_SPLIT_MLIR=OFF \
  -UIREE_COMPILER_PLUGIN_PATHS \
  -UIREE_EXTERNAL_HAL_DRIVERS \
  -DIREE_BUILD_EXPERIMENTAL_DISTRIBUTED=ON \
  -DIREE_ENABLE_COLLECTIVE_RUNTIME_TESTS=ON
```
Platform:
Tesla T4 GPU
CUDA version 12.3
NCCL version 2.19.3

**Comment by sogartar** [comment]
3b3cef90c895dc728cb57876d22b0fe108b39d70 indeed is the exact commit that introduces the problem.

**Comment by sogartar** [comment]
One problem that I encountered with address sanitizer is that the `iree_hal_cuda_pending_queue_actions_destroy` does not wait properly for all workers to finish executing the action queues. No one is waiting for the CUDA host stream callback `iree_hal_cuda_execution_device_signal_host_callback` to finish. We are only waiting for the worker to be idle (available to execute new actions), but the CUDA callback happens on another thread.
[asan.log](https://github.com/openxla/iree/files/14438210/asan.log)

**Comment by sogartar** [comment]
One way to fix this is to maintain a count of actions that are in execution and exit the worker loop only when this count is 0.

**Comment by antiagainst** [comment]
> One problem that I encountered with address sanitizer is that the `iree_hal_cuda_pending_queue_actions_destroy` does not wait properly for all workers to finish executing the action queues. No one is waiting for the CUDA host stream callback `iree_hal_cuda_execution_device_signal_host_callback` to finish. We are only waiting for the worker to be idle (available to execute new actions), but the CUDA callback happens on another thread. [asan.log](https://github.com/openxla/iree/files/14438210/asan.log)

Chatted offline. Yes this is the issue. We need to also wait for the kernel thread doing host func launch to finish. Idea is to add an atomic to track flying host func launches and wait for it to reach zero when existing the worker thread.

**Comment by benvanik** [comment]
(you'll want to make sure to use iree_notification_t to do the waiting)

**Comment by sogartar** [comment]
I think we have to notify [here](https://github.com/openxla/iree/blob/b2250d8ff4c81ab51095807944fe5c6809618ae0/runtime/src/iree/hal/drivers/cuda/pending_queue_actions.c#L549) after enqueuing more actions from the CUDA callback thread, that there is more work so that the loop in `iree_hal_cuda_worker_execute` does not wait forever [here](https://github.com/openxla/iree/blob/b2250d8ff4c81ab51095807944fe5c6809618ae0/runtime/src/iree/hal/drivers/cuda/pending_queue_actions.c#L878).
There is the possibility that [iree_hal_cuda_worker_process_ready_list](https://github.com/openxla/iree/blob/b2250d8ff4c81ab51095807944fe5c6809618ae0/runtime/src/iree/hal/drivers/cuda/pending_queue_actions.c#L903) has already exited before `iree_hal_cuda_execution_device_signal_host_callback` enqueues the new action.

I can add this in the PR I am preparing.

**Comment by sogartar** [comment]
Here are fixes for the concurrency issues #16621. I need to check on another machine that all is well. At least on one it works fine.
